staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/types"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/endpoints/request"
+
+	"github.com/google/uuid"
+)
+
+// If the length of an Audit-ID value exceeds the following threshold, we truncate it to keep
+// the first maxAuditIDLength characters.
+const maxAuditIDLength = 36
+
+// WithAuditID attaches the Audit-ID associated with a request to the context.
+//
+// a. If the caller does not specify a value for Audit-ID in the request header, we generate a new audit ID
+// b. If a user specified Audit-ID is too large, we truncate it.
+// c. We echo the Audit-ID value to the caller via the response Header 'Audit-ID'.
+func WithAuditID(handler http.Handler) http.Handler {
+	return withAuditID(handler, func() string {
+		return uuid.New().String()
+	})
+}
+
+func withAuditID(handler http.Handler, newAuditIDFunc func() string) http.Handler {
+	if newAuditIDFunc == nil {
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		var attachAuditIDToCtx bool
+		auditID := r.Header.Get(auditinternal.HeaderAuditID)
+		if len(auditID) == 0 {
+			attachAuditIDToCtx = true
+			auditID = newAuditIDFunc()
+		}
+
+		// if the user has specified a very long audit ID then we will use the first N characters
+		// the UUID is formatted as '47aa608a-8931-4573-bd42-e3e3bc912ff0' (36 chars)
+		// so let's ensure we retain at least the first 36 characters.
+		// Note: assuming Audit-ID header is in ASCII
+		if len(auditID) > 36 {
+			attachAuditIDToCtx = true
+			auditID = auditID[0:36]
+		}
```

Comment:
this is new. How do we know we don't break anybody? Why 36 chars?

Reply:
- yes, since we are copying the audit ID into the request context I wanted to a put a defensive guard to avoid copying large amount of data into the context. 
- Also, printing a very large audit ID is going to make reading the log very cumbersome. The client can still search with the first 36 characters.
- I chose 36, because the formatted UID we generate today is 36 characters long. I am fine increasing it but would still like us to have a limit.
- Well client-go does not generate any audit ID today. I don't think it breaks anyone today. The caller can still search with the first N characters.

Reply:
Does it mean the input hasn't been sanitized before? so clients could send potentially very large inputs.
AFAIK the HTTP spec doesn't specify a limit for Headers.

Reply:
for now, I am saving the audit ID as is in the context. But when we are logging (tracing and httplog) I am only showing the first N(64) character.